homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -259,10 +259,21 @@ def extra_authorize_data(self) -> dict:
         """
         return self.async_create_entry(title=self.flow_impl.name, data=data)
 
-    async_step_user = async_step_pick_implementation
-    async_step_ssdp = async_step_pick_implementation
-    async_step_zeroconf = async_step_pick_implementation
-    async_step_homekit = async_step_pick_implementation
+    async def async_step_discovery(self, user_input: dict = None) -> dict:
+        """Handle a flow initialized by discovery."""
+        await self.async_set_unique_id(self.DOMAIN)
+        self._abort_if_unique_id_configured()
+
+        assert self.hass is not None
+        if self.hass.config_entries.async_entries(self.DOMAIN):
```

Comment:
Why is this check needed? We abort above on same entries. Is it for backwards compatibility?

Reply:
The difference is, this will abort on "any" entry, the above will abort on the unique id.

Reply:
Ok, so it's for backwards compatibility.

Reply:
No, it is not.

The reason for this implementation is during a bit of play migrating the Spotify platform to an integration.

We can detect Spotify on the network (e.g., Spotify Connect-enabled devices, like a Sonos). This will tell us, the user is using Spotify. We can trigger auto-discovery on this for this integration: "Hey, we've found you using this, click to configure".

However, you won't be setting up the detected device, you are setting up a service. Hence, we need a way to stop discovery after the service has been set up.

This is not keeping backward compatibility, this is intended new logic.

Reply:
I understand it now. 

The unique ID is that we only ever configure a single integration via discovery. Since it's OAuth2, we can't know if something belongs to an account.

Reply:
Since this line catches all, we can remove the call to `_abort_if_unique_id_configured`, right?

Reply:
Yes.

Reply:
Nope, because we want to prevent triggering multiple discoveries on multiple devices.

Reply:
Where's that side effect happening by calling `_abort_if_unique_id_configured` and not happening when aborting here? I don't see it in the code.

Reply:
Setting the unique ID will abort for any config flow in progress with the same ID. `_abort_if_unique_id_configured` will only look at config entries (not flows). So it's superseded by the abort if we have any entry check below.

Reply:
Absolutely, you guys are right.
Opened PR: <https://github.com/home-assistant/home-assistant/pull/30733>